python/ribasim/ribasim/model.py

```diff
@@ -280,6 +285,107 @@ def write(self, filepath: str | PathLike[str]) -> Path:
         context_file_writing.set({})
         return fn
 
+    def _validate_model(self):
+        df_edge = self.edge.df
+        df_chunks = [node.node.df.set_crs(self.crs) for node in self._nodes()]
+        df_node = pd.concat(df_chunks)
+
+        df_graph = df_edge
+        # Join df_edge with df_node to get to_node_type
+        df_graph = df_graph.join(
+            df_node[["node_type"]], on="from_node_id", how="left", rsuffix="_from"
+        )
+        df_graph = df_graph.rename(columns={"node_type": "from_node_type"})
+
+        df_graph = df_graph.join(
+            df_node[["node_type"]], on="to_node_id", how="left", rsuffix="_to"
+        )
+        df_graph = df_graph.rename(columns={"node_type": "to_node_type"})
+
+        if not self._check_neighbors(
+            df_graph, flow_edge_amount, "flow", df_node["node_type"]
+        ):
+            raise ValueError("Minimum flow inneighbor or outneighbor unsatisfied")
+        if not self._check_neighbors(
+            df_graph, control_edge_amount, "control", df_node["node_type"]
+        ):
+            raise ValueError("Minimum control inneighbor or outneighbor unsatisfied")
+
+    def _check_neighbors(
+        self,
+        df_graph: pd.DataFrame,
+        edge_amount: dict[str, list[int]],
+        edge_type: str,
+        nodes,
+    ) -> bool:
+        is_valid = True
+        # Count edge neighbor
+        df_graph = df_graph.loc[df_graph["edge_type"] == edge_type]
+
+        # check from node's neighbor
+        from_node_count = (
+            df_graph.groupby("from_node_id").size().reset_index(name="from_node_count")  # type: ignore
+        )
+
+        df_result = (
+            df_graph[["from_node_id", "from_node_type"]]
+            .drop_duplicates()
+            .merge(from_node_count, on="from_node_id", how="left")
+        )
+
+        df_result = df_result[["from_node_id", "from_node_count", "from_node_type"]]
+        for index, node in enumerate(nodes):
+            if nodes.index[index] not in df_result["from_node_id"].to_numpy():
+                new_row = {
+                    "from_node_id": nodes.index[index],
+                    "from_node_count": 0,
+                    "from_node_type": node,
+                }
+                df_result = pd.concat(
+                    [df_result, pd.DataFrame([new_row])], ignore_index=True
+                )
+        for _, row in df_result.iterrows():
+            # from node's outneighbor
+            try:
+                if row["from_node_count"] < edge_amount[row["from_node_type"]][2]:
+                    is_valid = False
+                    raise ValueError(
```

Comment:
Why raise an error if you always catch it? You can just pass your message to logging.error?

Reply:
I wanted to give customized error message, I didn't know that logging.error() can already show the error message on the screen. I just checked and now I do know